lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/classify/MultiMADClassifier.java

```diff
@@ -0,0 +1,291 @@
+                System.arraycopy(metrics, 0, results, 0, metrics.length);
+                Arrays.parallelSetAll(results, m -> (
+                    (m > highCutoff && includeHigh) || (m < lowCutoff && includeLow)) ? 1.0 : 0.0);
+                output.addDoubleColumn(columnName + outputColumnSuffix, results);
+            } else {
+                // Non-parallel scoring:
+                double[] results = new double[metrics.length];
+                for (int r = 0; r < input.getNumRows(); r++) {
+                    boolean isOutlier = (metrics[r] > highCutoff && includeHigh) ||
+                        (metrics[r] < lowCutoff && includeLow);
+                    results[r] = isOutlier ? 1.0 : 0.0;
+                }
+                output.addDoubleColumn(columnName + outputColumnSuffix, results);
+            }
+        }
+    }
+
+    private int[] getRandomSample(int numRows) {
+        int sampleSize = (int)(numRows * samplingRate);
+        int[] sampleIndices = new int[sampleSize];
+
+        if (useReservoirSampling) {
+            // Reservoir Sampling:
+            for (int i = 0; i < sampleSize; i++) {
+                sampleIndices[i] = i;
+            }
+            Random rand = new Random();
+            for (int i = sampleSize; i < numRows; i++) {
+                int j = rand.nextInt(i+1);
+                if (j < sampleSize) {
+                    sampleIndices[j] = i;
+                }
+            }
+            boolean[] mask = new boolean[numRows];
+            for (int i = 0; i < sampleSize; i++) {
+                mask[sampleIndices[i]] = true;
+            }
+        } else {
+            // partial Fisher-Yates shuffle
+            int[] range = new int[numRows];
+            for (int i = 0; i < numRows; i++) {
+                range[i] = i;
+            }
+            Random rand = new Random();
+            for (int i = 0; i < sampleSize; i++) {
+                int j = rand.nextInt(numRows - i) + i;
+                int temp = range[j];
+                range[j] = range[i];
+                range[i] = temp;
+            }
+            System.arraycopy(range, 0, sampleIndices, 0, sampleSize);
+        }
+
+        return sampleIndices;
+    }
+
+    private Stats train(double[] metrics) {
+        double median = new Percentile().evaluate(metrics, 50);
+
+        for (int i = 0; i < metrics.length; i++) {
+            metrics[i] = Math.abs(metrics[i] - median);
+        }
+
+        double MAD = new Percentile().evaluate(metrics, 50);
+
+        if (MAD == 0) {
+            Arrays.sort(metrics);
+            int lowerTrimmedMeanIndex = (int) (metrics.length * trimmedMeanFallback);
+            int upperTrimmedMeanIndex = (int) (metrics.length * (1 - trimmedMeanFallback));
+            double sum = 0;
+            for (int i = lowerTrimmedMeanIndex; i < upperTrimmedMeanIndex; ++i) {
+                sum += metrics[i];
+            }
+            MAD = sum / (upperTrimmedMeanIndex - lowerTrimmedMeanIndex);
```

Comment:
Given that the point of MAD is to be robust to outliers, something like MAD = (p(10) + p(90)) / 2 wouldn't be good, because if p(90) is an outlier, then MAD is much higher than it should be. The current implementation finds the average of all residuals between the 5th and 95th percentiles, which is more robust, but I guess still not ideal.